File: cpp/lowest-common-ancestor-of-deepest-leaves.cpp
1123. Lowest Common Ancestor of Deepest Leaves
Medium

94

129

Favorite

Share
Given a rooted binary tree, return the lowest common ancestor of its deepest leaves.

Recall that:

The node of a binary tree is a leaf if and only if it has no children
The depth of the root of the tree is 0, and if the depth of a node is d, the depth of each of its children is d+1.
The lowest common ancestor of a set S of nodes is the node A with the largest depth such that every node in S is in the subtree with root A.
 

Example 1:

Input: root = [1,2,3]
Output: [1,2,3]
Explanation: 
The deepest leaves are the nodes with values 2 and 3.
The lowest common ancestor of these leaves is the node with value 1.
The answer returned is a TreeNode object (not an array) with serialization "[1,2,3]".
Example 2:

Input: root = [1,2,3,4]
Output: [4]
Example 3:

Input: root = [1,2,3,4,5]
Output: [2,4,5]
 

Constraints:

The given tree will have between 1 and 1000 nodes.
Each node of the tree will have a distinct value between 1 and 1000.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lcaDeepestLeaves(TreeNode* root) {
        auto res=dfs(root);
        return res.first;
    }
private:
    pair<TreeNode*,int> dfs(TreeNode* root) {
        if (root==NULL) return {NULL, 0};
        auto l=dfs(root->left), r=dfs(root->right);
        if (l.second==r.second) return {root, l.second+1};
        else {
            return {l.second>r.second?l.first:r.first, l.second>r.second?l.second+1:r.second+1};
        }
    }
};

File: python/lowest-common-ancestor-of-deepest-leaves.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def lcaDeepestLeaves(self, root: TreeNode) -> TreeNode:
        def dfs(root):
            if not root: return None, 0
            ln, lh=dfs(root.left)
            rn, rh=dfs(root.right)
            if lh==rh: return (root, lh+1)
            else: return (ln, lh+1) if lh>rh else (rn, rh+1)
        node, height=dfs(root)
        return node

